### PR TITLE
Optimize SNN training with in-place resets and Triton kernel expansion

### PR DESCRIPTION
**Pull request by fangwei123456**
## 概述

对 SpikingJelly 框架进行三项独立训练加速优化，互不依赖，均可合入 main：

1. **`reset()` 原地清零**：消除每轮训练的重复 GPU 内存分配
2. **Surrogate Triton 覆盖扩展**：修复 5 种 surrogate 在 triton 后端梯度为 0 的静默 bug，并扩展为 7 种
3. **LIF multi_step_forward 统一**：所有后端在 GPU 上自动路由到 Triton，`torch` 后端训练性能提升 29%

---

## 改动文件

```
spikingjelly/activation_based/base.py
spikingjelly/activation_based/triton_kernel/surrogate_kernel.py
spikingjelly/activation_based/neuron/lif.py
```

---

### 1. `base.py` — `MemoryModule.reset()` 原地清零

**问题**：每次调用 `reset_net(model)` 都通过 `copy.deepcopy` 分配新张量，
VGG-16-BN 有 13 个神经元层，即每轮训练 13 次 GPU 内存分配+释放。

**修复**：

```python
# 改前
self._memories[key] = copy.deepcopy(self._memories_rv[key])

# 改后（按情况选最优路径）
cur.detach_().copy_(rv)   # Tensor reset value：in-place，复用显存
cur.detach_().fill_(rv)   # Scalar reset value：in-place fill
copy.deepcopy(rv)         # 其他情况：fallback
```

`detach_()` 是关键：清除 stale `grad_fn`，防止 reset 后再 forward 时
触发"backward through graph a second time"错误。

---

### 2. `surrogate_kernel.py` — Triton surrogate 覆盖从 2 种扩展到 7 种

**问题**：`sg_triton()` 仅支持 Sigmoid（id=0）和 ATan（id=1），其余 surrogate
类型调用 `resolve_sg_triton_id_and_alpha()` 会抛 `NotImplementedError`，
导致使用其他 surrogate 的用户无法使用 `backend="triton"`。

**修复**：新增 5 种实现：

| id | Surrogate | Triton 表达式 |
|----|-----------|--------------|
| 2 | PiecewiseQuadratic | `max(0, alpha - alpha² * \|h\|)` |
| 3 | PiecewiseExp | `alpha/2 * exp(-alpha * \|h\|)` |
| 4 | SoftSign | `1 / (2α(1/α + \|h\|)²)` |
| 5 | SuperSpike | `alpha / (\|h\| + 1)²` |
| 6 | Erf | `alpha/√π * exp(-(αh)²)` |

所有超越函数使用 `tl.float32` 上转换以避免 fp16 精度问题。

---

### 3. `lif.py` — LIFNode multi_step_forward 统一

**问题**：
- eval 路径：`torch`/`cupy` 后端有 8 个独立的 `jit_eval_multi_step_forward_*` 方法（hard/soft reset × decay_input × store_v_seq），代码重复严重
- training 路径：`torch` 后端始终走 Python for 循环，即使 GPU 上有更快的 Triton 核

**事实澄清**：现有 `_multistep_lif_forward_kernel` 已通过 `tl.constexpr` 参数
覆盖所有 4 种变体（soft/hard reset × decay_input），无需重写内核。

**修复**：

**(a) training 路径 — `torch` 后端 GPU 自动路由到 Triton：**

```python
elif self.backend == "torch":
    if x_seq.is_cuda:
        self.v_float_to_tensor(x_seq[0])
        try:
            spike_seq, v_seq = triton_kernel.multistep_lif(...)
            ...
            return spike_seq
        except (NotImplementedError, AttributeError, TypeError):
            pass  # 不支持的 surrogate → Python loop
    return super().multi_step_forward(x_seq)
```

**(b) eval 路径 — 所有后端 GPU 统一走 Triton：**

```python
else:  # eval
    self.v_float_to_tensor(x_seq[0])
    if x_seq.is_cuda:
        try:
            spike_seq, v_seq = triton_kernel.multistep_lif(...)
            ...
            return spike_seq
        except (NotImplementedError, AttributeError, TypeError):
            pass
    # CPU 或不支持 surrogate：统一 fallback
    out = self._eval_multi_step_forward(
        x_seq, self.v, self.v_threshold, self.v_reset,
        self.tau, self.decay_input, self.store_v_seq,
    )
    ...
```

**(c) 新增 `_eval_multi_step_forward` 静态方法：**

单一 Python for 循环，覆盖所有 4 种变体，替代 8 个重复方法作为 CPU fallback。
原 8 个 `jit_eval_multi_step_forward_*` 方法保留（向后兼容，子类可能重写）。

---

### 正确性验证

3 种 surrogate × 2 种 reset × 2 种 decay × train/eval = **24 个配置**，
`torch` 后端与 `triton` 后端输出完全一致。

### 性能（RTX 4090，PyTorch 2.7.1）

**SpikingVGG-16-BN，T=4，B=64，CIFAR-10，前向+反向**：

| 后端 | ms/iter | img/s | 加速比 |
|------|---------|-------|-------|
| `torch`（改前，Python loop） | 38.93ms | 1644 | 1.00× |
| `torch`（改后，GPU→Triton） | 30.18ms | 2121 | **1.29×** |
| `triton`（参考） | 30.03ms | 2131 | 1.30× |

改后 `torch` 后端与 `triton` 后端性能差距 < 0.5%。

**孤立神经元层（T=32，B=64，N=512）**：

| 后端 | 改前 | 改后 | 加速 |
|------|------|------|------|
| `torch` training | ~23ms（Python loop） | ~1.6ms（Triton） | **14×** |
| `torch` eval | Python loop | Triton | 与 triton 后端相同 |

---

## API 兼容性

- **无破坏性变更**：`torch`/`triton`/`cupy` 后端接口不变
- `torch` 后端用户自动获得 Triton 加速，无需任何代码修改
- 使用不支持 surrogate（非 7 种之列）的用户自动 fallback 到 Python loop，
  并建议切换到 `backend="torch"`
- 8 个 `jit_eval_multi_step_forward_*` 方法保留，不影响现有子类

---

## 已知限制

- （稀疏脉冲 GPU 训练）：实测 dense cuBLAS 在典型 SNN 规模下全面优于稀疏方法，暂不可行
- `PiecewiseLeakyReLU` surrogate 有两个参数（`w` 和 `c`），当前 Triton 接口不支持，仍需 `backend="torch"`

## Summary by CodeRabbit

* **New Features**
  * Five new surrogate gradient options: PiecewiseQuadratic, PiecewiseExp, SoftSign, SuperSpike, Erf.

* **Performance**
  * Faster multi-step neuron inference with optimized GPU execution and a robust unified CPU fallback.
  * More efficient memory reset that avoids unnecessary copies for tensor and scalar memories.

* **Bug Fixes / Reliability**
  * Improved fallback behavior and clearer error guidance for unsupported surrogate types and kernel failures.

* **Tests**
  * New test ensuring GPU vs CPU inference produces matching results.

### CODE REVIEW

**Comment by fangwei123456** [comment]
@codex review

**Comment by fangwei123456** [comment]
@codex review

**Comment by fangwei123456** [comment]
@codex review

**Comment by fangwei123456** [comment]
@gemini-code-assist review

**Comment by fangwei123456** [comment]
@coderabbitai review
